### PR TITLE
feat: LAND Ping

### PR DESCRIPTION
**Pull request by fmiras**
### What?
LAND Ping is a `LANDRegistry` feature that allows users to "refresh" their last activity of their Assets. It defines a "grace period" for the feature to start running and It delegates the usage of a *decayed* LAND to another address (In this case for a future "Dead Assets Auction" smart contract)

### Why? 
There needs to be a way for LAND parcels that were lost due to human error, neglect or abandonment to be put into productive use. Not doing so would deter the experience of the virtual world.

### CODE REVIEW

File: contracts/land/ILANDRegistry.sol [hunk 1]
```diff
@@ -59,4 +65,15 @@ interface ILANDRegistry {
...
+  event Ping(
+    address indexed _caller,
+    address indexed _target
+  );
+
+  event Reactivate(
+    uint256 indexed _assetId,
+    address indexed _newOwner
+  );
+
```
Comment (nachomazzara) [inline contracts/land/ILANDRegistry.sol]: What do you think of adding an event when someone set a new ReactivateAuthorized?

```solidity
event ReactivateAuthorized(address indexed _caller, address indexed _operator, bool _authorized)
```

File: contracts/land/ILANDRegistry.sol [hunk 2]
```diff
@@ -59,4 +65,15 @@ interface ILANDRegistry {
     address indexed _caller,
     address indexed _deployer
   );
+
+  event Ping(
+    address indexed _caller,
+    address indexed _target
```
Comment (nachomazzara) [inline contracts/land/ILANDRegistry.sol]: I would call it `_holder` as the other events we have.

File: contracts/land/LANDRegistry.sol [hunk 3]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  //
+  // LAND Ping
+  //
+
+  /**
+   * @dev Set date from when the LAND Ping feature should be enabled
+   * @param _gracePeriod Desired amount of time in miliseconds to enable feature
+   */
+  function setGracePeriod(uint _gracePeriod) external onlyDeployer {
+    gracePeriod = _gracePeriod;
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I'm thinking if we should add a `require` here to check that the grace period has been set to ensure that won't be updated.

Anyways the contract is upgradable.

What do you think @eordano @abarmat?

File: contracts/land/LANDRegistry.sol [hunk 4]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  }
+
+  /**
+   * @dev Set amount of time that should pass for a LAND to be transferred to
+   * a new onwer
+   * @param _deemPeriod Desired amount of time in miliseconds for a LAND to decay
+   */
+  function setDeemPeriod(uint _deemPeriod) external onlyDeployer {
+    deemPeriod = _deemPeriod;
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: same as above @eordano @abarmat maybe this one has more sense. The grace period may change due the time could take to develop the auction feature, but changing this one could be polemic

File: contracts/land/LANDRegistry.sol [hunk 5]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  }
+
+  /**
+   * @dev Ping an address, keep all that address LAND "alive"
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I would change `Ping an address, keep all that address LAND "alive"` to 
`Ping an address`. 

What do you think?

File: contracts/land/LANDRegistry.sol [hunk 6]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+   */
+  function ping(address _user) external {
+    require(
+      _isApprovedForAll(_user, msg.sender) ||
+      updateManager[_user][msg.sender],
+      "This function can only be called by the operatorForAll or updateManager"
+    );
+    // solium-disable-next-line security/no-block-members
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Is this disable necessary?

File: contracts/land/LANDRegistry.sol [hunk 7]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  }
+
+  function _ping(address _address) private {
+    // solium-disable-next-line security/no-block-members
+    latestPing[_address] = block.timestamp;
+    emit Ping(msg.sender, _address);
+  }
+
+  /**
+   * @dev Check if LAND is still alive or not
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I think that should be: Check if `a` LAND is still alive or not

Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: Doc param is missing

File: contracts/land/LANDRegistry.sol [hunk 8]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    _ping(_user);
+  }
+
+  /**
+   * @dev Ping myself, keep all my LAND "alive". This is only refresh owned assets, not operated/allowed ones
+   */
+  function pingMyself() external {
+    _ping(msg.sender);
+  }
+
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Dev notation is missing here

File: contracts/land/LANDRegistry.sol [hunk 9]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    // solium-disable-next-line security/no-block-members
+    latestPing[_address] = block.timestamp;
+    emit Ping(msg.sender, _address);
+  }
+
+  /**
+   * @dev Check if LAND is still alive or not
+   */
+  function hasDecayed(uint256 _tokenId) external view returns (bool) {
+    return _hasDecayed(_tokenId);
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I'm wondering why it is calling an internal function. Is there any reason?

File: contracts/land/LANDRegistry.sol [hunk 10]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  }
+
+  /**
+   * @dev Ping myself, keep all my LAND "alive". This is only refresh owned assets, not operated/allowed ones
+   */
+  function pingMyself() external {
+    _ping(msg.sender);
+  }
+
+  function _ping(address _address) private {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I would suggest to use `internal`. We are not expecting to use this contract as `parent` of another one

File: contracts/land/LANDRegistry.sol [hunk 11]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+
+  //
+  // LAND Ping
+  //
+
+  /**
+   * @dev Set date from when the LAND Ping feature should be enabled
+   * @param _gracePeriod Desired amount of time in miliseconds to enable feature
+   */
+  function setGracePeriod(uint _gracePeriod) external onlyDeployer {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Use `uint256`

File: contracts/land/LANDRegistry.sol (hunk 4 above)
Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: Use `uint256`

File: contracts/land/LANDRegistry.sol [hunk 12]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    }
+
+    // solium-disable-next-line security/no-block-members
+    if (block.timestamp < gracePeriod) {
+      return false;
+    }
+
+    address owner = _ownerOf(_tokenId);
+    // solium-disable-next-line security/no-block-members
+    return latestPing[owner] + deemPeriod < block.timestamp;
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Warning!!

We can have an [arithmetic overflow](https://medium.com/coinmonks/ethernaut-lvl-5-walkthrough-how-to-abuse-arithmetic-underflows-and-overflows-2c614fa86b74) here.

Adding the following check can help to reduce the risk.:
```Solidity
uint256 lastPing = latestPing[owner] + deemPeriod
require(lastPing >= deemPeriod, "Overflow");
```

Maybe we are not going to live to see it happen but will prevent any kind of undesired behavior

File: contracts/land/LANDRegistry.sol [hunk 13]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  }
+
+  /**
+   * @dev Check if LAND is still alive or not
+   */
+  function hasDecayed(uint256 _tokenId) external view returns (bool) {
+    return _hasDecayed(_tokenId);
+  }
+
+  function _hasDecayed(uint256 _tokenId) private view returns (bool) {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I think that this abstraction not worth it, but if we still need it I would suggest using `internal` instead of `private`

File: contracts/land/LANDRegistry.sol [hunk 14]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    emit Ping(msg.sender, _address);
+  }
+
+  /**
+   * @dev Check if LAND is still alive or not
+   */
+  function hasDecayed(uint256 _tokenId) external view returns (bool) {
+    return _hasDecayed(_tokenId);
+  }
+
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Dev notation is missing here

File: contracts/land/LANDRegistry.sol [hunk 15]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  function _hasDecayed(uint256 _tokenId) private view returns (bool) {
+    if (gracePeriod == 0) {
+      return false;
+    }
+
+    // solium-disable-next-line security/no-block-members
+    if (block.timestamp < gracePeriod) {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I would make it inclusive to match the requirements `block.timestamp <= gracePeriod` 

If the grace period is the same as the block timestamp we are still good.

File: contracts/land/LANDRegistry.sol (hunk 12 above)
Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: Also, I would make it inclusive as above

File: contracts/land/LANDRegistry.sol [hunk 16]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    }
+
+    address owner = _ownerOf(_tokenId);
+    // solium-disable-next-line security/no-block-members
+    return latestPing[owner] + deemPeriod < block.timestamp;
+  }
+
+  /**
+   * @dev Authorize address to manage decayed assets (Auction Contract)
+   */
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Param doc missing here.

File: contracts/land/LANDRegistry.sol [hunk 17]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  function setReactivateAuthorized(address _address, bool _isAuthorized) external onlyDeployer {
+    reactivateAuthorized[_address] = _isAuthorized;
+  }
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: I would emit an event here.

File: contracts/land/LANDRegistry.sol [hunk 18]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+
+  function reactivate(uint256 _tokenId, address _newOwner) external onlyReactivateAuthorized {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Missing dev notation

File: contracts/land/LANDRegistry.sol [hunk 19]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    require(_hasDecayed(_tokenId), "LAND has not decayed, can not reactivate");
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Let's keep this method empty. It won't consume was if someone call if by mistake

Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: Hmm or maybe we can just remove it.

CC @eordano @abarmat

File: test/LANDRegistry.js [hunk 20]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
       await assertRevert(land.assignNewParcel(0, 3, user, sentByOperator))
     })
   })
+
+  describe('ping', () => {
+    it('should set reactivate authorized', async () => {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: I would have a separate `describe` for:

`setReactivateAuthorized`
`ping`
`setGracePeriod`
`setDeemPeriod`

File: contracts/land/LANDRegistry.sol [hunk 21]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
     return size > 0;
   }
+
+  //
+  // LAND Ping
+  //
+
+  /**
+   * @dev Set date from when the LAND Ping feature should be enabled
+   * @param _gracePeriod Desired amount of time in miliseconds to enable feature
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Based on how you are using it on `hasDecayed`, it seems to be a specific date in seconds.

Solidity `block.timestamp` is in `seconds`.

File: contracts/land/LANDRegistry.sol [hunk 22]
```diff
@@ -479,4 +472,87 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+   */
+  function setGracePeriod(uint _gracePeriod) external onlyDeployer {
+    gracePeriod = _gracePeriod;
+  }
+
+  /**
+   * @dev Set amount of time that should pass for a LAND to be transferred to
+   * a new onwer
+   * @param _deemPeriod Desired amount of time in miliseconds for a LAND to decay
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: It should be in seconds by the comment I left above

File: test/LANDRegistry.js [hunk 23]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
...
+
+    it('should set deem period', async () => {
+      let deemPeriod = await land.deemPeriod()
+      deemPeriod.should.be.bignumber.equal(0)
+      await land.setDeemPeriod(1, sentByCreator)
+      deemPeriod = await land.deemPeriod()
+      deemPeriod.should.be.bignumber.equal(1)
+    })
+
+    it('should return true if decayed LAND', async () => {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: Maybe we can change  `should return true if decayed LAND` to `should return true if a LAND is decayed`

File: test/LANDRegistry.js (hunk 20 above)
Comment (nachomazzara) [reply test/LANDRegistry.js]: I think that we are using 
```javascript
async function () {}
``` 
everywhere. It would be great if you can follow that

File: test/LANDRegistry.js [hunk 24]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
...
+      const assetId = await land.encodeTokenId(0, 1)
+      await land.setGracePeriod(1, sentByCreator)
+      await land.setDeemPeriod(2, sentByCreator)
+      const decayed = await land.hasDecayed(assetId)
+      decayed.should.be.true
+    })
+
+    it('should return false if no gracePeriod set', async () => {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: `should return false if no gracePeriod has been set`

File: test/LANDRegistry.js [hunk 25]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
...
+      const assetId = await land.encodeTokenId(0, 1)
+      const decayed = await land.hasDecayed(assetId)
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: Tip: you can use `1` as `assetId`. The `assetId` for every LAND following the coordinates pattern `(0,x)`  is `x`

File: test/LANDRegistry.js [hunk 26]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
...
+
+    it('should return false if gracePeriod is bigger than today', async () => {
+      const assetId = await land.encodeTokenId(0, 1)
+      const now = new Date().getMilliseconds() / 1000
+      await land.setGracePeriod(now, sentByCreator)
+      const decayed = await land.hasDecayed(assetId)
+      decayed.should.be.false
+    })
+
+    it('should return false if not decayed LAND', async () => {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: `should return false if the LAND has not decayed`

File: test/LANDRegistry.js [hunk 27]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
...
+      const assetId = await land.encodeTokenId(0, 1)
+      const now = new Date().getMilliseconds() * 1000
+      await land.setGracePeriod(1, sentByCreator)
+      await land.setDeemPeriod(now * now, sentByCreator)
+      const decayed = await land.hasDecayed(assetId)
+      decayed.should.be.false
+    })
+
+    it('should reactivate decayed LAND and emit reactivate event', async () => {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: It would be better to remove this test until we have the functionality done

File: test/LANDRegistry.js [hunk 28]
```diff
@@ -1098,4 +1097,150 @@ contract('LANDRegistry', accounts => {
...
+      await land.reactivate(assetId, anotherUser, sentByOperator)
+      const latestPingAfter = await land.latestPing(anotherUser)
+      ;(latestPingBefore < latestPingAfter).should.be.true
+    })
+
+    it('should refresh latestPing if pinged by owner', async () => {
+      const latestPingBefore = await land.latestPing(user)
+      await land.pingMyself(sentByUser)
+      const latestPingAfter = await land.latestPing(user)
+      ;(latestPingBefore < latestPingAfter).should.be.true
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: What do you think of using a variable to store it? It will make the test more readable.

Also, it would be great if you can include the event emission check.

Comment (nachomazzara) [reply test/LANDRegistry.js]: Please, try checking against the specific time of the ping, it will make the test solid

File: contracts/land/LANDRegistry.sol (hunk 5 above)
Comment (fmiras) [reply contracts/land/LANDRegistry.sol]: It's fine by me

File: test/LANDRegistry.js [hunk 29]
```diff
@@ -1098,4 +1099,176 @@ contract('LANDRegistry', accounts => {
...
+
+      it('should return false if gracePeriod is bigger than today (one month from now)', async function() {
+        const assetId = await land.encodeTokenId(0, 1)
+        const date = new Date()
+        const oneMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0)
+          .getDate() * 24 * 60 * 60 // In seconds
+
+        await land.setGracePeriod(oneMonth, sentByCreator)
+        const decayed = await land.hasDecayed(assetId)
+        decayed.should.be.false
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: It could be returning false because the deeming period is not set. Do you mind checking it too?

File: test/LANDRegistry.js [hunk 30]
```diff
@@ -1098,4 +1099,176 @@ contract('LANDRegistry', accounts => {
...
+      it('should emit Ping event when ping by other', async function() {
+        await land.setApprovalForAll(anotherUser, true, sentByUser)
+        const { logs } = await land.ping(user, sentByAnotherUser)
+        const log = logs[0]
+        log.event.should.be.eq('Ping')
+        log.args._caller.should.be.bignumber.equal(anotherUser)
+        log.args._holder.should.be.equal(user)
+      })
+
+      it('reverts if ping non-authorized address', async function() {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: `reverts if ping non-authorized address` by `reverts if trying to ping by a non-authorized address`

File: contracts/land/LANDRegistry.sol (hunk 4 above)
Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: Can we emit an event here too?

File: contracts/land/LANDRegistry.sol [hunk 31]
```diff
@@ -479,4 +464,78 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  // LAND Ping
+  //
+
+  /**
+   * @dev Set date from when the LAND Ping feature should be enabled
+   * @param _gracePeriod Desired amount of time in seconds from now to enable feature
+   */
+  function setGracePeriod(uint256 _gracePeriod) external onlyDeployer {
+    // solium-disable-next-line security/no-block-members
+    gracePeriod = block.timestamp + _gracePeriod;
```
[the remark this answers is not in the document]
Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: Can we emit an event here too?

File: test/LANDRegistry.js [hunk 32]
```diff
@@ -1098,4 +1099,176 @@ contract('LANDRegistry', accounts => {
...
   })
+
+  describe('LAND ping', function() {
+    describe('setGracePeriod', function() {
+      it('should set 1 month grace period', async function() {
+        let gracePeriod = await land.gracePeriod()
+        gracePeriod.should.be.bignumber.equal(0)
+        
+        const date = new Date()
+        const oneMonth = new Date(date.getFullYear(), date.getMonth() + 1, 0)
```
[the remark this answers is not in the document]
Comment (fmiras) [reply test/LANDRegistry.js]: Because not every month represent the same amount of seconds, It could be 30, 31, 28 or 29 days . An the test is using today related dates

[the remark this answers is not in the document]
Comment (fmiras) [reply test/LANDRegistry.js]: That's why I'm using `new Date(date.getFullYear(), date.getMonth() + 1, 0).getDate()` instead of `31`

[the remark this answers is not in the document]
Comment (nachomazzara) [reply test/LANDRegistry.js]: I know, it is just an avg to make the test more semantic. We shouldn't care about the days of the month, but to respect the input set

File: test/LANDRegistry.js [hunk 33]
```diff
@@ -1098,4 +1099,176 @@ contract('LANDRegistry', accounts => {
...
+        const dp = await land.deemPeriod()
+        await sleep(3) // inside delta period
+        const decayed = await land.hasDecayed(assetId)
+        decayed.should.be.false
+      })
+
+
+      it('should return false if no gracePeriod set', async function() {
+        const assetId = await land.encodeTokenId(0, 1)
+        const decayed = await land.hasDecayed(assetId)
```
[the remark this answers is not in the document]
Comment (fmiras) [reply test/LANDRegistry.js]: What do you mean? How do you think that it must be tested?

[the remark this answers is not in the document]
Comment (nachomazzara) [reply test/LANDRegistry.js]: The test says that the `gracePeriod` is not set so:

```javascript
const gracePeriod = await land.gracePreiod()
gracePeriod.should.be.bignumber.equal(0)
```

File: contracts/land/LANDRegistry.sol [hunk 34]
```diff
@@ -488,6 +488,7 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
   function setDeemPeriod(uint256 _deemPeriod) external onlyDeployer {
     require(_deemPeriod != 0, "Deem period can not be 0");
     deemPeriod = _deemPeriod;
+    emit DeemPeriod(deemPeriod);
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Please, can we add the caller?

File: contracts/land/LANDRegistry.sol [hunk 35]
```diff
@@ -479,4 +464,78 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    emit DeemPeriod(msg.sender, deemPeriod);
+  }
+
+  /**
+   * @dev Ping an address
+   * @param _user address of LAND holder to be pinged
+   */
+  function ping(address _user) external {
+    require(
+      _isApprovedForAll(_user, msg.sender) ||
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: As we talked offline. Let's add the ability for the `proxyOwner` to update it too

File: contracts/land/LANDRegistry.sol [hunk 36]
```diff
@@ -479,4 +464,78 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+
+  //
+  // LAND Ping
+  //
+
+  /**
+   * @dev Set the date from when the LAND Ping feature should be enabled
+   * @param _gracePeriod Desired amount of time in seconds from now to enable the feature
+   */
+  function setGracePeriod(uint256 _gracePeriod) external onlyDeployer {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Lets use `onlyProxyOwner` here

File: contracts/land/LANDRegistry.sol [hunk 37]
```diff
@@ -479,4 +464,78 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+    gracePeriod = block.timestamp.add(_gracePeriod);
+    emit GracePeriod(msg.sender, gracePeriod);
+  }
+
+  /**
+   * @dev Set the amount of time that should pass for a LAND to be transferred to
+   * a new onwer
+   * @param _deemPeriod Desired amount of time in seconds for a LAND to decay
+   */
+  function setDeemPeriod(uint256 _deemPeriod) external onlyDeployer {
```
Comment (nachomazzara) [inline contracts/land/LANDRegistry.sol]: Lets use `onlyProxyOwner` here

File: contracts/land/LANDRegistry.sol [hunk 38]
```diff
@@ -479,4 +464,78 @@ contract LANDRegistry is Storage, Ownable, FullAssetRegistry, ILANDRegistry {
...
+  function _ping(address _address) internal {
+    // solium-disable-next-line security/no-block-members
+    latestPing[_address] = block.timestamp;
+    emit Ping(msg.sender, _address);
+  }
+
+  /**
+   * @dev Check if a LAND is still alive or not
+   * @param _assetId LAND encoded coordinates
+   */
```
[the remark this answers is not in the document]
Comment (nachomazzara) [reply contracts/land/LANDRegistry.sol]: It should be `@return boolean - whether the LAND is decayed or not`

File: test/LANDRegistry.js [hunk 39]
```diff
@@ -1098,4 +1099,176 @@ contract('LANDRegistry', accounts => {
       await assertRevert(land.assignNewParcel(0, 3, user, sentByOperator))
     })
   })
+
+  describe('LAND ping', function() {
```
[the remark this answers is not in the document]
Comment (nachomazzara) [reply test/LANDRegistry.js]: Pending

File: test/LANDRegistry.js [hunk 40]
```diff
@@ -1098,4 +1103,168 @@ contract('LANDRegistry', accounts => {
...
+    describe('setGracePeriod', function() {
+      it('should set 4 weeks grace period', async function() {
+        let gracePeriod = await land.gracePeriod()
+        gracePeriod.should.be.bignumber.equal(0)
+        await land.setGracePeriod(duration.weeks(4), sentByCreator)
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: `duration.weeks(4)` can be stored in a variable. It is being used in a lot of places.

File: test/LANDRegistry.js [hunk 41]
```diff
@@ -1098,4 +1103,168 @@ contract('LANDRegistry', accounts => {
...
+        await increaseTime(duration.weeks(5)) // outside delta period
+        const decayed = await land.hasDecayed(assetId)
+        decayed.should.be.true
+      })
+
+      it('should return false is the LAND is not decayed but near to deem period', async function() {
+        const assetId = await land.encodeTokenId(0, 1)
+        await land.setGracePeriod(duration.weeks(1), sentByCreator)
+        await land.setDeemPeriod(duration.weeks(4), sentByCreator)
+        await increaseTime(duration.weeks(2)) // inside delta period
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: I would add also the case of `4 weeks and 6 days and 59 minutes` (edge case)

File: test/LANDRegistry.js [hunk 42]
```diff
@@ -1098,4 +1103,168 @@ contract('LANDRegistry', accounts => {
...
+      })
+
+      it('should return false if no deemPeriod set', async function() {
+        const deemPeriod = await land.deemPeriod()
+        deemPeriod.should.be.bignumber.equal(0)
+        const assetId = await land.encodeTokenId(0, 1)
+        const decayed = await land.hasDecayed(assetId)
+        decayed.should.be.false
+      })
+
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: What happens if both grace and deem periods are not set?

File: test/LANDRegistry.js (hunk 29 above)
Comment (nachomazzara) [reply test/LANDRegistry.js]: Pending

File: test/LANDRegistry.js [hunk 43]
```diff
@@ -1098,4 +1103,168 @@ contract('LANDRegistry', accounts => {
...
+        const log = logs[0]
+        log.event.should.be.eq('Ping')
+        log.args._caller.should.be.bignumber.equal(anotherUser)
+        log.args._holder.should.be.equal(user)
+      })
+
+      it('reverts if trying to ping by a non-authorized address', async function() {
+        await assertRevert(land.ping(user, sentByHacker))
+      })
+    })
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: We should add the new test about being pinged by the `proxyOwner`

File: test/LANDRegistry.js [hunk 44]
```diff
@@ -1182,12 +1183,21 @@ contract('LANDRegistry', accounts => {
...
         const assetId = await land.encodeTokenId(0, 1)
         await land.setGracePeriod(duration.weeks(1), sentByCreator)
-        await land.setDeemPeriod(duration.weeks(4), sentByCreator)
+        await land.setDeemPeriod(fourWeeksDuration, sentByCreator)
         await increaseTime(duration.weeks(2)) // inside delta period
         const decayed = await land.hasDecayed(assetId)
         decayed.should.be.false
       })
 
+      it.only('should return false is the LAND is not decayed (edge case)', async function() {
```
Comment (nachomazzara) [inline test/LANDRegistry.js]: Ups
